File: 695-max-area-of-island/695-max-area-of-island.rs
impl Solution {
    pub fn max_area_of_island(mut grid: Vec<Vec<i32>>) -> i32 {
        let (m, n) = (grid.len(), grid[0].len());
        let mut area = 0;
        
        for i in 0..m{
            for j in 0..n{
                if grid[i][j] == 1{
                    area = area.max(dfs(&mut grid, i, j));
                }
            }
        }
        
        area
    }
}

fn dfs(grid: &mut Vec<Vec<i32>>, i: usize, j: usize) -> i32{
    let (m, n) = (grid.len(), grid[0].len());
    
    grid[i][j] = 0;
    
    let mut sum = 1;

    for (i, j) in [(i+1, j), (i-1, j), (i, j+1), (i, j-1)]{
        if i != usize::MAX && j != usize::MAX && i < m && j < n && grid[i][j] == 1{
            sum += dfs(grid, i, j);
        }
    }
    
    sum
}

File: 695-max-area-of-island/695-max-area-of-island.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> bool:
        def dfs(i: int, j: int)-> int:
            sum = 0
            if i >= 0 and j >= 0 and i < m and j < n:
                if grid[i][j] == 1:
                    grid[i][j] = 0
                    sum = 1
                    for (dx, dy) in [(1 ,0), (0, 1), (-1, 0), (0, -1)]:
                        x = i + dx
                        y = j + dy
                        sum+=dfs(x, y)

            if sum >=1:
                print(i, j)
            return sum
                    
        m, n = len(grid), len(grid[0])
        cnt = 0
        for i in range(0, m):
            for j in range(0, n):
                if grid[i][j] == 1:
                    cnt = max(dfs(i, j), cnt)
                    print("---------")
                    
        return cnt